File: tests/arr_test.cpp
#include "gtest/gtest.h"
#include "../arr.h"
#include "../arr_serialize.h"
#include "../serialize.h"
#include <string>
#include <sstream>
#include <fstream>
#include <cstdio>

using namespace std;

class OutputCapture {
    stringstream buffer;
    streambuf* prev;
public:
    OutputCapture() : prev(cout.rdbuf(buffer.rdbuf())) {}
    ~OutputCapture() { cout.rdbuf(prev); }
    string str() { return buffer.str(); }
};

// проверка на управление памятью при расширении и сжатии массива
TEST(MyArrTest, Branch_CapacityAndShrink) {
    MyArr arr;

    for(int i=0; i<65; ++i) {
        arr.addEnd("x");
    }
    EXPECT_GE(arr.lenArr(), 65);
    
    int startSize = arr.lenArr();
    for(int i=0; i < startSize; ++i) {
        arr.delEnd(); 
    }
    EXPECT_EQ(arr.lenArr(), 0);
    
    for(int i=0; i<20; ++i) arr.addEnd("y");
    for(int i=0; i<20; ++i) arr.delHead();
    EXPECT_EQ(arr.lenArr(), 0);

    for(int i=0; i<20; ++i) arr.addEnd("z");
    for(int i=0; i<20; ++i) arr.delAt(0);
    EXPECT_EQ(arr.lenArr(), 0);
}

// проверка на граничные случаи и обработку ошибок индексов
TEST(MyArrTest, Branch_EdgeCases) {
    MyArr arr;
    OutputCapture cap; 

    // Операции с пустым массивом
    arr.readArray();
    arr.delHead();
    arr.delEnd();
    arr.delAt(0);
    arr.repArr(0, "val");
    
    // Неверные индексы 
    arr.addEnd("A");
    arr.addAt(-1, "err");
    arr.addAt(5, "err");
    arr.delAt(-1);
    arr.delAt(5);
    arr.repArr(-1, "err");
    arr.repArr(5, "err");
    
    EXPECT_EQ(arr.getAt(-1), "[INVALID_INDEX]");
    EXPECT_EQ(arr.getAt(5), "[INVALID_INDEX]");

    arr.addHead("Head");
    EXPECT_EQ(arr.getAt(0), "Head");
    
    // Вставка  середину
    arr.addAt(1, "Mid"); 
    EXPECT_EQ(arr.getAt(1), "Mid");

    EXPECT_NE(arr.getData_test(), nullptr);
}

// проверка на обработку ошибок ввода-вывода
TEST(MyArrTest, IO_Errors_Coverage) {
    MyArr arr;
    
    ArrSerializer::loadFromFile(arr, "no_such_file.arr");
    ArrSerializer::loadFromBinaryFile(arr, "no_such_file.bin");

    {
        ofstream f("corrupt.arr");
        f << "NotANumber"; 
    }
    ArrSerializer::loadFromFile(arr, "corrupt.arr");
    remove("corrupt.arr");

    {
        ofstream f("corrupt.bin", ios::binary);
        int size = 100; 
        f.write((char*)&size, sizeof(size));
        // данные не пишем
    }
    ArrSerializer::loadFromBinaryFile(arr, "corrupt.bin");
    remove("corrupt.bin");
    

    ArrSerializer::saveToFile(arr, ""); 
    ArrSerializer::saveToBinaryFile(arr, "");
}

// проверка на корректную работу сериализации и десериализации
TEST(MyArrTest, SaveAndLoad_HappyPath) {
    MyArr arr;
    arr.addEnd("1");
    arr.addEnd("2");
    
    ArrSerializer::saveToFile(arr, "good.arr");
    MyArr arr2;
    ArrSerializer::loadFromFile(arr2, "good.arr");
    EXPECT_EQ(arr2.lenArr(), 2);
    EXPECT_EQ(arr2.getAt(0), "1");
    remove("good.arr");

    ArrSerializer::saveToBinaryFile(arr, "good.bin");
    MyArr arr3;
    ArrSerializer::loadFromBinaryFile(arr3, "good.bin");
    EXPECT_EQ(arr3.lenArr(), 2);
    remove("good.bin");
}

File: ds/arr_test.go
package ds

import (
	"os"
	"testing"
)

func TestMyArr_Branch_CapacityAndShrink(t *testing.T) {
	arr := NewMyArr()
	for i := 0; i < 65; i++ {
		arr.AddEnd("x")
	}
	if arr.LenArr() < 65 {
		t.Errorf("Expected size >= 65, got %d", arr.LenArr())
	}
	startSize := arr.LenArr()
	for i := 0; i < startSize; i++ {
		arr.DelEnd()
	}
	if arr.LenArr() != 0 {
		t.Errorf("Expected empty array, got size %d", arr.LenArr())
	}
	for i := 0; i < 20; i++ {
		arr.AddEnd("y")
	}
	for i := 0; i < 20; i++ {
		arr.DelHead()
	}
	if arr.LenArr() != 0 {
		t.Errorf("Expected empty after DelHead loop")
	}
	for i := 0; i < 20; i++ {
		arr.AddEnd("z")
	}
	for i := 0; i < 20; i++ {
		arr.DelAt(0)
	}
	if arr.LenArr() != 0 {
		t.Errorf("Expected empty after DelAt loop")
	}
}

func TestMyArr_Branch_EdgeCases(t *testing.T) {
	arr := NewMyArr()
	arr.ReadArray()
	arr.DelHead()
	arr.DelEnd()
	arr.DelAt(0)
	arr.RepArr(0, "val")

	arr.AddEnd("A")
	arr.AddAt(-1, "err")
	arr.AddAt(5, "err")
	arr.DelAt(-1)
	arr.DelAt(5)
	arr.RepArr(-1, "err")
	arr.RepArr(5, "err")

	if val := arr.GetAt(-1); val != "[INVALID_INDEX]" {
		t.Errorf("Expected [INVALID_INDEX], got %s", val)
	}
	if val := arr.GetAt(5); val != "[INVALID_INDEX]" {
		t.Errorf("Expected [INVALID_INDEX], got %s", val)
	}

	arr.AddHead("Head")
	arr.AddAt(1, "Mid")
	if arr.GetAt(0) != "Head" {
		t.Errorf("Expected Head at 0")
	}
	if arr.GetAt(1) != "Mid" {
		t.Errorf("Expected Mid at 1")
	}
	
	arr.AddAt(arr.LenArr(), "TailAppend")
	if arr.GetAt(arr.LenArr()-1) != "TailAppend" {
		t.Error("AddAt at end failed")
	}
}

func TestMyArr_SaveAndLoad(t *testing.T) {
	arr := NewMyArr()
	arr.AddEnd("one")
	arr.AddEnd("two")
	filename := "arr_test.bin"
	
	serializer := NewArrSerializer()
	if err := serializer.SaveToFile(arr, filename); err != nil {
		t.Fatalf("Save failed: %v", err)
	}
	defer os.Remove(filename)

	arr2 := NewMyArr()
	if err := serializer.LoadFromFile(arr2, filename); err != nil {
		t.Fatalf("Load failed: %v", err)
	}

	if arr2.LenArr() != 2 {
		t.Errorf("Size mismatch: expected 2, got %d", arr2.LenArr())
	}
	if arr2.GetAt(0) != "one" || arr2.GetAt(1) != "two" {
		t.Errorf("Content mismatch")
	}
}	

func TestMyArr_FullCoverage(t *testing.T) {
	arr := NewMyArr()

	arr.AddAt(0, "A")     
	arr.AddAt(5, "B")   
	arr.AddAt(-1, "C")    
	arr.AddAt(1, "Tail")   
	arr.AddAt(0, "Head")  
	
	if arr.GetAt(0) != "Head" || arr.GetAt(2) != "Tail" {
		t.Error("AddAt logic failed")
	}

	arr.DelAt(-1) 
	arr.DelAt(100) 
	arr.DelAt(1)
	if arr.LenArr() != 2 { t.Error("DelAt failed") }

	arr.RepArr(-1, "X")
	arr.RepArr(100, "X")
	arr.RepArr(0, "NewHead")
	if arr.GetAt(0) != "NewHead" { t.Error("RepArr failed") }

	if arr.GetAt(-5) != "[INVALID_INDEX]" { t.Error("GetAt negative failed") }
	if arr.GetAt(100) != "[INVALID_INDEX]" { t.Error("GetAt overflow failed") }

	empty := NewMyArr()
	empty.DelHead()
	empty.DelEnd()  
	empty.ReadArray() 

	serializer := NewArrSerializer()
	serializer.SaveToFile(arr, "arr_cov.txt")
	defer os.Remove("arr_cov.txt")
	arr2 := NewMyArr()
	serializer.LoadFromFile(arr2, "arr_cov.txt")
	if arr2.LenArr() != arr.LenArr() { t.Error("Text Save/Load size mismatch") }

	serializer.SaveToBinaryFile(arr, "arr_cov.bin")
	defer os.Remove("arr_cov.bin")
	arr3 := NewMyArr()
	serializer.LoadFromBinaryFile(arr3, "arr_cov.bin")
	if arr3.GetAt(0) != "NewHead" { t.Error("Binary Save/Load val mismatch") }
}